File: app/schemas/solicitacao.py
from typing import Optional, List, Dict, Any
from datetime import date, datetime
from decimal import Decimal
from pydantic import BaseModel, Field, validator
from app.schemas.nota_fiscal import NotaFiscalProcessadaOut
from app.constants import TIPOS_PLANILHA_LEGADO
import os

class NotaIgnoradaOut(BaseModel):
    numero_nota: str
    serie: Optional[str] = None
    chave_acesso: Optional[str] = None
    data_emissao: Optional[str] = None
    motivo: str
    arquivo: Optional[str] = None


class ItemExcluidoOut(BaseModel):
    chave_acesso: Optional[str] = None
    numero_nota: str
    serie: Optional[str] = None
    item_numero: int
    arquivo: Optional[str] = None
    destino: str
    ncm: str
    descricao: str
    descricao_confiavel: bool = True
    motivo: str
    tipo_exclusao: str
    regras_aplicadas: Optional[List[Dict[str, Any]]] = Field(default_factory=list)
    v_total: Optional[Decimal] = None
    base_calculo: Optional[Decimal] = None
    ipi_despesas: Optional[Decimal] = None
    a_ori: Optional[Decimal] = None
    a_dst: Optional[Decimal] = None
    debito: Optional[Decimal] = None
    credito: Optional[Decimal] = None
    valor_devido: Optional[Decimal] = None


class AvisoAvaliacaoOut(BaseModel):
    numero_nota: str
    serie: Optional[str] = None
    item_numero: int
    arquivo: Optional[str] = None
    aviso: str


class SolicitacaoCreate(BaseModel):
    empresa_id: int = Field(..., example=1)
    periodo_inicio: date = Field(..., example="2026-01-01")
    periodo_fim: date = Field(..., example="2026-01-31")
    tipo_planilha: Optional[str] = Field(
        None,
        example="multi",
        description="Legado: antecipacao_parcial, antecipacao_tributaria ou difal. Se omitido, o sistema roteia automaticamente por CFOP e gera todas as planilhas aplicáveis."
    )
    template_id: Optional[int] = Field(None, description="Legado: usado apenas junto com tipo_planilha para o fluxo de tipo único")

    @validator("periodo_fim")
    def validate_periodo(cls, v, values):
        if "periodo_inicio" in values and v < values["periodo_inicio"]:
            raise ValueError("periodo_fim não pode ser anterior a periodo_inicio")
        return v

    @validator("tipo_planilha")
    def validate_tipo_planilha(cls, v):
        if v is None:
            return v
        if v not in TIPOS_PLANILHA_LEGADO:
            raise ValueError(
                f"tipo_planilha deve ser um dos seguintes: {', '.join(TIPOS_PLANILHA_LEGADO)}"
            )
        return v

    @validator("template_id")
    def validate_template_requires_tipo(cls, value, values):
        if value is not None and not values.get("tipo_planilha"):
            raise ValueError("template_id só pode ser usado junto com tipo_planilha")
        return value

    class Config:
        extra = "forbid"

class SolicitacaoSaidaOut(BaseModel):
    id: str
    solicitacao_id: str
    tipo: str
    template_id: Optional[int]
    arquivo_path: Optional[str]
    total_notas: int
    total_valor_devido: Decimal
    aviso: Optional[str]
    criado_em: datetime

    @validator("id", "solicitacao_id", pre=True)
    def ensure_str_uuid(cls, v):
        return str(v) if v is not None else v

    @validator("arquivo_path", pre=True)
    def hide_internal_path(cls, value):
        return os.path.basename(str(value).replace("\\", "/")) if value else value

    class Config:
        orm_mode = True

class SolicitacaoOut(BaseModel):
    id: str
    empresa_id: int
    usuario_id: Optional[str] = None
    periodo_inicio: date
    periodo_fim: date
    tipo_planilha: str
    template_id: Optional[int]
    status: str
    mensagem_erro: Optional[str]
    arquivo_saida_path: Optional[str]
    total_notas_processadas: int
    notas_ignoradas: Optional[List[Dict[str, Any]]] = Field(default_factory=list)
    itens_excluidos: Optional[List[Dict[str, Any]]] = Field(default_factory=list)
    avisos_avaliacao: Optional[List[Dict[str, Any]]] = Field(default_factory=list)
    cfops_sem_regra: Optional[Dict[str, int]] = Field(default_factory=dict)
    criado_em: datetime
    atualizado_em: datetime
    notas_processadas: Optional[List[NotaFiscalProcessadaOut]] = Field(default_factory=list)
    saidas: Optional[List[SolicitacaoSaidaOut]] = Field(default_factory=list)

    @validator("id", "usuario_id", pre=True)
    def ensure_str_uuid(cls, v):
        return str(v) if v is not None else v

    @validator("arquivo_saida_path", pre=True)
    def hide_internal_path(cls, value):
        return os.path.basename(str(value).replace("\\", "/")) if value else value

    class Config:
        orm_mode = True


class SolicitacaoListOut(BaseModel):
    id: str
    empresa_id: int
    usuario_id: Optional[str] = None
    periodo_inicio: date
    periodo_fim: date
    tipo_planilha: str
    template_id: Optional[int]
    status: str
    mensagem_erro: Optional[str]
    total_notas_processadas: int
    criado_em: datetime
    atualizado_em: datetime

    @validator("id", "usuario_id", pre=True)
    def ensure_str_uuid(cls, value):
        return str(value) if value is not None else value

    class Config:
        orm_mode = True

File: frontend/src/types/solicitacao.ts
import type { JsonObject } from './common';

export interface NotaFiscalProcessada {
  id: string;
  solicitacao_id: string;
  chave_acesso?: string | null;
  numero_nota: string;
  serie?: string | null;
  cnpj_emitente?: string | null;
  uf_emitente?: string | null;
  cnpj_destinatario: string;
  uf_destinatario?: string | null;
  data_emissao: string;
  data_entrada?: string | null;
  origem_data_entrada?: 'planilha_sistema_contabil' | 'sped_fiscal' | 'xml_nfe' | 'manual' | null;
  item_numero: number;
  ncm: string;
  cfop?: string | null;
  destino_planilha?: TipoPlanilha | null;
  v_total: number;
  base_calculo: number;
  ipi_despesas: number;
  a_ori: number;
  a_dst_resolvida: number;
  debito: number;
  credito: number;
  valor_devido: number;
  metadados_extras: JsonObject;
  criado_em: string;
}

export interface NotaIgnorada {
  numero_nota: string;
  serie?: string | null;
  chave_acesso?: string | null;
  data_emissao?: string | null;
  motivo: string;
  arquivo?: string | null;
}

export interface ItemExcluido {
  chave_acesso?: string | null;
  numero_nota: string;
  serie?: string | null;
  item_numero: number;
  arquivo?: string | null;
  destino: string;
  ncm: string;
  descricao: string;
  descricao_confiavel: boolean;
  motivo: string;
  tipo_exclusao: 'mercadoria' | 'aliquotas_iguais';
  regras_aplicadas?: JsonObject[];
  v_total?: number | null;
  base_calculo?: number | null;
  ipi_despesas?: number | null;
  a_ori?: number | null;
  a_dst?: number | null;
  debito?: number | null;
  credito?: number | null;
  valor_devido?: number | null;
}

export interface AvisoAvaliacao {
  numero_nota: string;
  serie?: string | null;
  item_numero: number;
  arquivo?: string | null;
  aviso: string;
}

export type TipoPlanilha =
  | 'antecipacao_parcial'
  | 'antecipacao_parcial_antecipado'
  | 'antecipacao_parcial_simples'
  | 'antecipacao_parcial_antecipado_simples'
  | 'antecipacao_tributaria'
  | 'antecipacao_tributaria_antecipado'
  | 'difal';
export type StatusSolicitacao = 'pendente' | 'processando' | 'concluido' | 'erro';

export interface SolicitacaoSaida {
  id: string;
  solicitacao_id: string;
  tipo: TipoPlanilha;
  template_id?: number | null;
  arquivo_path?: string | null;
  total_notas: number;
  total_valor_devido: number;
  aviso?: string | null;
  criado_em: string;
}

export interface Solicitacao {
  id: string;
  empresa_id: number;
  periodo_inicio: string;
  periodo_fim: string;
  tipo_planilha: TipoPlanilha | 'multi';
  template_id?: number | null;
  status: StatusSolicitacao;
  mensagem_erro?: string | null;
  arquivo_saida_path?: string | null;
  total_notas_processadas: number;
  notas_ignoradas?: NotaIgnorada[];
  itens_excluidos?: ItemExcluido[];
  avisos_avaliacao?: AvisoAvaliacao[];
  cfops_sem_regra?: Record<string, number>;
  criado_em: string;
  atualizado_em: string;
  notas_processadas?: NotaFiscalProcessada[];
  saidas?: SolicitacaoSaida[];
}

export interface SolicitacaoCreate {
  empresa_id: number;
  periodo_inicio: string;
  periodo_fim: string;
  tipo_planilha?: TipoPlanilha;
  template_id?: number;
}

export interface NotaBonificacaoPendencia {
  chave_acesso: string;
  numero_nota: string;
  serie?: string;
  cnpj_emitente?: string;
  nome_emitente?: string;
  cfops: string[];
  valor_total: number;
  tem_credito: boolean;
  sugestao_revenda: boolean;
  motivo_sugestao: string;
}

export interface PreAnaliseSolicitacao {
  requer_decisao: boolean;
  notas_bonificacao: NotaBonificacaoPendencia[];
}

export interface SolicitacoesBatchDeleteRequest {
  ids: string[];
}

export interface SolicitacoesBatchDeleteResponse {
  deleted_count: number;
  ids: string[];
}
